File: V2rayNG/library/src/main/java/com/neko/marquee/text/AutoMarqueeTextView.java
package com.neko.marquee.text;

import android.content.Context;
import android.util.AttributeSet;
import androidx.appcompat.widget.AppCompatTextView;

public class AutoMarqueeTextView extends AppCompatTextView {
    @Override // android.view.View
    public boolean isFocused() {
        return true;
    }

    public AutoMarqueeTextView(Context context) {
        super(context);
    }

    public AutoMarqueeTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AutoMarqueeTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }
}

File: V2rayNG/library/src/main/java/com/neko/marquee/text/AutoMarqueeTextView.kt
package com.neko.marquee.text

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import androidx.appcompat.widget.AppCompatTextView

class AutoMarqueeTextView : AppCompatTextView {
    private var mAggregatedVisible: Boolean = false

    constructor(context: Context) : super(context) {
        mAggregatedVisible = false
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        mAggregatedVisible = false
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        mAggregatedVisible = false
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        isSelected = true
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        isSelected = false
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        onVisibilityAggregated(isVisibleToUser())
    }

    override fun onVisibilityAggregated(isVisible: Boolean) {
        super.onVisibilityAggregated(isVisible)
        if (isVisible == mAggregatedVisible) {
            return
        }
        mAggregatedVisible = isVisible
        if (mAggregatedVisible) {
            ellipsize = TextUtils.TruncateAt.MARQUEE
        } else {
            ellipsize = TextUtils.TruncateAt.END
        }
    }
    
    fun View.isVisibleToUser(): Boolean {
        return this.visibility == View.VISIBLE
    }
}
